### PR TITLE
Riemann fit speed up

### PR DESCRIPTION
**Pull request by rovere**
The Riemann fit has been reworked and divided into smaller kernels, and most of the variables have been moved to global memory. This allowed reducing by almost a factor 2 the registers required by the heaviest kernel and allowed more "concurrent" executions.

Other improvements:

* avoid branching on barrel-forward case, where possible
* move all to double and have full reproducibility between the CPU and GPU fit results
* fix the stand-alone test
* Refactor function names
* Tune the number of block and thread/block taking into account the reduced number of required registers

**NOTE**: it may be necessary to boost the stack size to 32Kb. I did it locally, but not modified the runTheMatrix workflows.

### CODE REVIEW

**Comment by fwyzard** [comment]
I have tried increasing the stack size up to 64k, but the jobs always fail with
```
Begin processing the 1st record. Run 321177, Event 187588887, LumiSection 142 on stream 0 at 06-Oct-2018 02:55:37.568 CEST
/mnt/home/fwyzard/patatrack/CMSSW_10_2_5_Patatrack/src/RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu, line 443:
cudaErrorLaunchOutOfResources: too many resources requested for launch
```

I am running on data, with this PR on top of the current HEAD.

**Comment by rovere** [comment]
ouch, that's bad. Of course, I had tested this PR on TTBar+50PU, but I see you are running on real data.
I can only suggest reducing the number of blocks in a few places.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu [hunk 1]
```diff
@@ -360,21 +426,39 @@ void CAHitQuadrupletGeneratorGPU::launchKernels(const TrackingRegion &region,
...
+  numberOfBlocks = (maxNumberOfQuadruplets_ + blockSize - 1) / blockSize;
+
+  kernelFastFitAllHits<<<numberOfBlocks, blockSize, 0, cudaStream>>>(
       d_foundNtupletsVec_[regionIndex], hh.gpu_d, 4, bField_, helix_fit_resultsGPU_,
       hitsGPU_, hits_covGPU_, circle_fit_resultsGPU_, fast_fit_resultsGPU_,
       line_fit_resultsGPU_);
   cudaCheck(cudaGetLastError());
 
-  blockSize = 256;
+  blockSize = 768;
```
Comment (rovere) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu]: This is a good candidate to lower back to 256

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu [hunk 2]
```diff
@@ -330,7 +396,7 @@ void CAHitQuadrupletGeneratorGPU::launchKernels(const TrackingRegion &region,
 
   auto nhits = hh.nHits;
   assert(nhits <= PixelGPUConstants::maxNumberOfHits);
-  auto blockSize = 64;
+  auto blockSize = 128;
```
Comment (rovere) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu]: This, most likely, is either a mistake of mine or I got this in from other commits.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu [hunk 3]
```diff
@@ -360,21 +426,39 @@ void CAHitQuadrupletGeneratorGPU::launchKernels(const TrackingRegion &region,
 
   // kernel_print_found_ntuplets<<<1, 1, 0, cudaStream>>>(d_foundNtupletsVec_[regionIndex], 10);
 
-  kernelFastFitAllHits<<<numberOfBlocks, 512, 0, cudaStream>>>(
+  blockSize = 768;
```
Comment (rovere) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu]: Also this one could be lowered back to 256

**Comment by rovere** [comment]
One last comment could be useful: I tested everything with 2 streams, 8 threads. I could be that increasing the number of streams would require too many resources.
Maybe we should agree on a fixed set of workflows and configuration to test **before** submitting the PR. I do not like to waste people's time.

**Comment by rovere** [comment]
In any case,
```
runTheMatrix.py -l 136.8648
```
worked with this PR on vinavx2 w/o issues.

**Comment by rovere** [comment]
I confirm  I can run also on the data pointed to by Andreas, w/o any issue and even w/o increasing the stack-size. I run with 2 streams and 8 threads.
The only issue I encountered after processing 6K+ events is a failed assert that I report here for reference:

```
cmsRun: /data/user/fwyzard/patatrack/build/slc7_amd64_gcc700.patatrack/tmp/BUILDROOT/add72b6485998bbd191bb064b1c1a622/opt/cmssw/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_2_5_Patatrack/src/RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitHeterogeneous.cc:245: void SiPixelRecHitHeterogeneous::run(const edm::Handle<edmNew::DetSetVector<SiPixelCluster> >&, SiPixelRecHitCollectionNew&, const HitsOnCPU&) const: Assertion `ngh==DSViter->size()' failed.
```

**Comment by fwyzard** [comment]
Here is the comparison of the performance on a V100 before and after this PR:
plot

**Comment by rovere** [comment]
So I should at least change the name of the branch .... ;)

**Comment by fwyzard** [comment]
@rovere , what should we do about this PR ?

**Comment by rovere** [comment]
Close it.

**Pull request closed without merging**